unexpected `Unknown column` error in downstream SQL execution

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

Running a workload contains parallel DDL(drop column) and DML(insert rows) from upstream. Kill drainer process and restart it periodically.

2. What did you expect to see?

All txn synced to downstream successfully.

3. What did you see instead?

Unexpected SQL exec failure in downstream.

drainer's log, it seems drainer use the old schema when restarting.

```
[2020/12/30 06:43:27.220 +00:00] [ERROR] [executor.go:111] ["Exec fail, will rollback"] [query="REPLACE INTO `test0`.`t_column`(`col_0`,`col_1`,`col_10`,`col_100`,`col_101`,`col_102`,`col_103`,`col_104`,`col_105`,`col_106`,`col_107`,`col_108`,`col_109`,`col_11`,`col_110`,`col_111`,`col_112`,`col_113`,`col_114`,`col_115`,`col_116`,`col_117`,`col_118`,`col_119`,`col_12`,`col_120`,`col_121`,`col_122`,`col_123`,`col_124`,`col_125`,`col_126`,`col_127`,`col_128`,`col_129`,`col_13`,`col_130`,`col_131`,`col_132`,`col_133`,`col_134`,`col_135`,`col_136`,`col_137`,`col_138`,`col_139`,`col_14`,`col_140`,`col_141`,`col_142`,`col_143`,`col_144`,`col_145`,`col_146`,`col_147`,`col_148`,`col_149`,`col_15`,`col_150`,`col_151`,`col_152`,`col_153`,`col_154`,`col_155`,`col_156`,`col_157`,`col_158`,`col_159`,`col_16`,`col_160`,`col_161`,`col_162`,`col_163`,`col_164`,`col_165`,`col_166`,`col_167`,`col_168`,`col_169`,`col_17`,`col_170`,`col_171`,`col_172`,`col_173`,`col_174`,`col_175`,`col_176`,`col_177`,`col_178`,`col_179`,`col_18`,`col_180`,`col_181`,`col_182`,`col_183`,`col_184`,`col_185`,`col_186`,`col_187`,`col_188`,`col_189`,`col_19`,`col_190`,`col_191`,`col_192`,`col_2`,`col_20`,`col_21`,`col_22`,`col_23`,`col_24`,`col_25`,`col_26`,`col_27`,`col_28`,`col_29`,`col_3`,`col_30`,`col_31`,`col_32`,`col_33`,`col_34`,`col_35`,`col_36`,`col_37`,`col_38`,`col_39`,`col_4`,`col_40`,`col_41`,`col_42`,`col_43`,`col_44`,`col_45`,`col_46`,`col_47`,`col_48`,`col_49`,`col_5`,`col_50`,`col_51`,`col_52`,`col_53`,`col_54`,`col_55`,`col_56`,`col_57`,`col_58`,`col_59`,`col_6`,`col_60`,`col_61`,`col_62`,`col_63`,`col_64`,`col_65`,`col_66`,`col_67`,`col_68`,`col_69`,`col_7`,`col_70`,`col_71`,`col_72`,`col_73`,`col_74`,`col_75`,`col_76`,`col_77`,`col_78`,`col_79`,`col_8`,`col_80`,`col_81`,`col_82`,`col_83`,`col_84`,`col_85`,`col_86`,`col_87`,`col_88`,`col_89`,`col_9`,`col_90`,`col_91`,`col_92`,`col_93`,`col_94`,`col_95`,`col_96`,`col_97`,`col_98`,`col_99`) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"] [args="[\"compassionate_nightingale2\",\"kind_swirles0\",\"3950-07-25\",\"2004-07-21 09:34:26\",-90,98,\"6961-04-03\",\"2027-08-03 13:50:12\",\"blissful_kilby7\",\"1971-04-09 08:40:21\",\"nostalgic_heyrovsky4\",5301552180275134678,\"1084-05-18\",\"6260-11-20\",\"sad_lamarr8\",\"9819-01-28 23:50:07\",\"upbeat_kowalevski4\",-1791152217,\"confident_wilson8\",-84,\"2002-11-14 07:43:39\",-44,119,-94,\"xenodochial_davinci2\",\"4877-06-08 17:49:37\",\"nervous_spence3\",38,\"9051-09-16 16:43:15\",-5180950427725031972,3602985627705997711,-1072116896,2467811505727114967,\"2031-02-27 15:43:39\",\"4539-02-24\",-1859618694,\"1972-09-24 09:41:00\",\"vibrant_carson2\",\"elastic_elion3\",38,4233923991649445888,\"1542-07-10 22:59:25\",\"2027-11-16 18:44:38\",\"1977-10-02 17:14:21\",\"2017-08-26 09:43:57\",\"hardcore_keller4\",\"6513-03-03 03:34:07\",\"6986-09-22\",\"8950-06-12\",20,\"frosty_nobel7\",\"pensive_mclean5\",\"2425-05-19 01:03:09\",3520640029451418058,2760962328665949392,\"9318-10-23 10:35:51\",\"gracious_fermat7\",\"2020-07-15 04:40:54\",\"1991-11-23 21:24:39\",1075507185,\"2025-11-03 23:32:41\",-15,-2442826236682808865,\"youthful_curie3\",\"3318-11-19 23:02:04\",81,\"5042-08-30\",175390384484264623,95,522460201,91,1656879407,\"sleepy_perlman9\",\"2009-03-02 07:35:26\",\"2011-08-16 00:13:48\",\"naughty_khorana3\",\"awesome_nobel2\",\"2035-06-29 09:54:51\",-88,-2090008838,\"1972-06-14 11:52:56\",\"kickass_visvesvaraya0\",-1599989022337279681,110,\"2075-07-28\",-732812558,\"agitated_keller2\",1928278110,-1001097229,-695725022,\"romantic_shockley9\",\"2033-08-03 09:03:48\",23,\"9171-05-23 09:12:46\",2030246438,\"1981-10-17 17:48:26\",\"gracious_keller0\",\"8571-08-08\",-1987857342,-120,-1111383485,\"angry_brattain0\",\"hopeful_darwin8\",-3730743407709889854,989591683,-2857903376183074731,\"1496-09-11 02:02:41\",\"5209-10-19\",\"2019-12-17 18:19:48\",\"1971-07-08 15:46:52\",\"3184-03-12\",\"7698-12-31 17:00:31\",\"4093-12-29\",83,-176637065214860353,\"vibrant_hawking8\",\"7817-12-27\",\"sleepy_mclean0\",\"5563-03-20 05:41:22\",-4088915639649370902,-23,\"1987-10-21 12:30:45\",\"1974-08-21 18:09:40\",\"jovial_noyce2\",63,\"5851-05-24 03:04:59\",366122435773933690,1002315853,-64,\"sharp_wozniak6\",\"dreamy_kowalevski5\",-81,\"1986-06-03 03:42:41\",1544761196157362222,\"competent_jones9\",-800441093493433333,\"3517-06-12\",\"distracted_babbage9\",-1,\"4924-03-02\",\"5259-10-16\",\"7543-04-29 20:58:56\",\"2014-02-05 12:19:18\",\"3665-01-15 16:25:11\",\"zen_brahmagupta0\",\"7455-08-30\",\"blissful_payne3\",128863639,\"adoring_bartik5\",\"2034-11-13 18:04:50\",-62,-98,-112,-4728566752613062867,124,\"lucid_albattani0\",\"confident_saha4\",\"flamboyant_jepsen5\",1727662135,\"amazing_kowalevski4\",\"2020-03-21 01:31:35\",-9,\"8023-08-05 10:10:24\",4420639666106874084,\"quirky_carson2\",\"1984-07-27 12:54:15\",-490132347624135250,591844220,\"1988-04-13 21:28:42\",3577539075605726659,\"nostalgic_chandrasekhar9\",\"7502-08-09 03:02:04\",\"gracious_austin8\",\"7254-09-20\",\"frosty_shockley9\",-1212200628,\"2013-03-07 08:54:27\",\"1973-06-19 08:38:13\",\"epic_minsky9\",75,-128333474,\"6727-06-06 12:51:04\",31,\"8240-08-03 14:59:28\",-74,\"2017-09-10 16:48:18\",\"7584-04-02\",\"1989-10-01 21:12:29\",-1044965856,\"wizardly_agnesi8\",\"peaceful_edison0\",\"2024-09-03 00:00:19\",\"2013-10-21 18:22:01\"]"] [error="Error 1054: Unknown column 'col_100' in 'field list'"]
```

4. Please provide the relate downstream type and version of drainer.
(run `drainer -V` in terminal to get drainer's version)

```
Release Version: v4.0.9
Git Commit Hash: 3475d287e4da3bd872b90d5386a9918f6fbd7b72
Build TS: 2020-12-19 04:47:12
Go Version: go1.13
Go OS/Arch: linux/amd64
```

--- a/drainer/checkpoint/checkpoint.go
+++ b/drainer/checkpoint/checkpoint.go
@@ -31,10 +31,13 @@ type CheckPoint interface {
 	Load() error
 
 	// Save saves checkpoint information.
-	Save(commitTS int64, secondaryTS int64, consistent bool) error
+	Save(commitTS int64, secondaryTS int64, consistent bool, version int64) error
 
 	// TS gets checkpoint commit timestamp.
 	TS() int64
+
+	// SchemaVersion gets checkpoint current schemaversion.
+	SchemaVersion() int64
 
 	// IsConsistent return the Consistent status saved.
 	IsConsistent() bool

--- a/drainer/checkpoint/checkpoint.go
+++ b/drainer/checkpoint/checkpoint.go
@@ -64,7 +64,7 @@ func NewCheckPoint(cfg *Config) (CheckPoint, error) {
 		return nil, errors.Annotatef(err, "initialize %s type checkpoint with config %+v", cfg.CheckpointType, cfg)
 	}
 
-	log.Info("initialize checkpoint", zap.String("type", cfg.CheckpointType), zap.Int64("checkpoint", cp.TS()), zap.Reflect("cfg", cfg))
+	log.Info("initialize checkpoint", zap.String("type", cfg.CheckpointType), zap.Int64("checkpoint", cp.TS()), zap.Int64("version", cp.SchemaVersion()), zap.Reflect("cfg", cfg))
 
 	return cp, nil
 }

--- a/drainer/checkpoint/file.go
+++ b/drainer/checkpoint/file.go
@@ -93,7 +93,9 @@ func (sp *FileCheckPoint) Save(ts, secondaryTS int64, consistent bool, version i
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
-	sp.Version = version
+	if version > sp.Versoin {
+		sp.Version = version
+	}
 
 	var buf bytes.Buffer
 	e := toml.NewEncoder(&buf)

--- a/drainer/checkpoint/file.go
+++ b/drainer/checkpoint/file.go
@@ -93,7 +93,7 @@ func (sp *FileCheckPoint) Save(ts, secondaryTS int64, consistent bool, version i
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
-	if version > sp.Versoin {
+	if version > sp.Version {
 		sp.Version = version
 	}
 

--- a/drainer/checkpoint/file.go
+++ b/drainer/checkpoint/file.go
@@ -33,7 +33,7 @@ type FileCheckPoint struct {
 
 	ConsistentSaved bool  `toml:"consistent" json:"consistent"`
 	CommitTS        int64 `toml:"commitTS" json:"commitTS"`
-	Version 	    int64   `toml:"schema-version" json:"schema-version"`
+	Version         int64 `toml:"schema-version" json:"schema-version"`
 }
 
 // NewFile creates a new FileCheckpoint.

--- a/drainer/checkpoint/file.go
+++ b/drainer/checkpoint/file.go
@@ -33,6 +33,7 @@ type FileCheckPoint struct {
 
 	ConsistentSaved bool  `toml:"consistent" json:"consistent"`
 	CommitTS        int64 `toml:"commitTS" json:"commitTS"`
+	Version 	    int64   `toml:"schema-version" json:"schema-version"`
 }
 
 // NewFile creates a new FileCheckpoint.
@@ -82,7 +83,7 @@ func (sp *FileCheckPoint) Load() error {
 }
 
 // Save implements CheckPoint.Save interface
-func (sp *FileCheckPoint) Save(ts, secondaryTS int64, consistent bool) error {
+func (sp *FileCheckPoint) Save(ts, secondaryTS int64, consistent bool, version int64) error {
 	sp.Lock()
 	defer sp.Unlock()
 
@@ -92,6 +93,7 @@ func (sp *FileCheckPoint) Save(ts, secondaryTS int64, consistent bool) error {
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
+	sp.Version = version
 
 	var buf bytes.Buffer
 	e := toml.NewEncoder(&buf)
@@ -114,6 +116,14 @@ func (sp *FileCheckPoint) TS() int64 {
 	defer sp.RUnlock()
 
 	return sp.CommitTS
+}
+
+// SchemaVersion implements CheckPoint.SchemaVersion interface.
+func (sp *FileCheckPoint) SchemaVersion() int64 {
+	sp.RLock()
+	defer sp.RUnlock()
+
+	return sp.Version
 }
 
 // IsConsistent implements CheckPoint interface

--- a/drainer/checkpoint/file_test.go
+++ b/drainer/checkpoint/file_test.go
@@ -33,7 +33,7 @@ func (t *testCheckPointSuite) TestFile(c *C) {
 
 	testTs := int64(1)
 	// save ts
-	err = meta.Save(testTs, 0, false)
+	err = meta.Save(testTs, 0, false, 0)
 	c.Assert(err, IsNil)
 	// check ts
 	ts := meta.TS()
@@ -41,7 +41,7 @@ func (t *testCheckPointSuite) TestFile(c *C) {
 	c.Assert(meta.IsConsistent(), Equals, false)
 
 	// check consistent true case.
-	err = meta.Save(testTs, 0, true)
+	err = meta.Save(testTs, 0, true, 0)
 	c.Assert(err, IsNil)
 	ts = meta.TS()
 	c.Assert(ts, Equals, testTs)
@@ -70,6 +70,6 @@ func (t *testCheckPointSuite) TestFile(c *C) {
 	err = meta.Close()
 	c.Assert(err, IsNil)
 	c.Assert(errors.Cause(meta.Load()), Equals, ErrCheckPointClosed)
-	c.Assert(errors.Cause(meta.Save(0, 0, true)), Equals, ErrCheckPointClosed)
+	c.Assert(errors.Cause(meta.Save(0, 0, true, 0)), Equals, ErrCheckPointClosed)
 	c.Assert(errors.Cause(meta.Close()), Equals, ErrCheckPointClosed)
 }

--- a/drainer/checkpoint/mysql.go
+++ b/drainer/checkpoint/mysql.go
@@ -41,6 +41,7 @@ type MysqlCheckPoint struct {
 	ConsistentSaved bool             `toml:"consistent" json:"consistent"`
 	CommitTS        int64            `toml:"commitTS" json:"commitTS"`
 	TsMap           map[string]int64 `toml:"ts-map" json:"ts-map"`
+	Version         int64              `toml:"schema-version" json:"schema-version"`
 }
 
 var _ CheckPoint = &MysqlCheckPoint{}
@@ -126,7 +127,7 @@ func (sp *MysqlCheckPoint) Load() error {
 }
 
 // Save implements checkpoint.Save interface
-func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool) error {
+func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool, version int64) error {
 	sp.Lock()
 	defer sp.Unlock()
 
@@ -136,6 +137,7 @@ func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool) error {
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
+	sp.Version = version
 
 	if secondaryTS > 0 {
 		sp.TsMap["primary-ts"] = ts
@@ -170,6 +172,13 @@ func (sp *MysqlCheckPoint) TS() int64 {
 	defer sp.RUnlock()
 
 	return sp.CommitTS
+}
+// SchemaVersion implements CheckPoint.SchemaVersion interface.
+func (sp *MysqlCheckPoint) SchemaVersion() int64 {
+	sp.RLock()
+	defer sp.RUnlock()
+
+	return sp.Version
 }
 
 // Close implements CheckPoint.Close interface

--- a/drainer/checkpoint/mysql.go
+++ b/drainer/checkpoint/mysql.go
@@ -41,7 +41,7 @@ type MysqlCheckPoint struct {
 	ConsistentSaved bool             `toml:"consistent" json:"consistent"`
 	CommitTS        int64            `toml:"commitTS" json:"commitTS"`
 	TsMap           map[string]int64 `toml:"ts-map" json:"ts-map"`
-	Version         int64              `toml:"schema-version" json:"schema-version"`
+	Version         int64            `toml:"schema-version" json:"schema-version"`
 }
 
 var _ CheckPoint = &MysqlCheckPoint{}
@@ -173,6 +173,7 @@ func (sp *MysqlCheckPoint) TS() int64 {
 
 	return sp.CommitTS
 }
+
 // SchemaVersion implements CheckPoint.SchemaVersion interface.
 func (sp *MysqlCheckPoint) SchemaVersion() int64 {
 	sp.RLock()

--- a/drainer/checkpoint/mysql.go
+++ b/drainer/checkpoint/mysql.go
@@ -137,7 +137,7 @@ func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool, version 
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
-	if version > sp.Versoin {
+	if version > sp.Version {
 		sp.Version = version
 	}
 

--- a/drainer/checkpoint/mysql.go
+++ b/drainer/checkpoint/mysql.go
@@ -137,7 +137,9 @@ func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool, version 
 
 	sp.CommitTS = ts
 	sp.ConsistentSaved = consistent
-	sp.Version = version
+	if version > sp.Versoin {
+		sp.Version = version
+	}
 
 	if secondaryTS > 0 {
 		sp.TsMap["primary-ts"] = ts

--- a/drainer/checkpoint/mysql.go
+++ b/drainer/checkpoint/mysql.go
@@ -14,9 +14,11 @@
 package checkpoint
 
 import (
+	"context"
 	"database/sql"
 	"encoding/json"
 	"sync"
+	"time"
 
 	"github.com/pingcap/errors"
 	"github.com/pingcap/log"
@@ -25,6 +27,7 @@ import (
 	// mysql driver
 	_ "github.com/go-sql-driver/mysql"
 	"github.com/pingcap/tidb-binlog/pkg/loader"
+	"github.com/pingcap/tidb-binlog/pkg/util"
 )
 
 // MysqlCheckPoint is a local savepoint struct for mysql
@@ -152,12 +155,13 @@ func (sp *MysqlCheckPoint) Save(ts, secondaryTS int64, consistent bool, version 
 	}
 
 	sql := genReplaceSQL(sp, string(b))
-	_, err = sp.db.Exec(sql)
-	if err != nil {
-		return errors.Annotatef(err, "query sql failed: %s", sql)
-	}
-
-	return nil
+	return util.RetryContext(context.TODO(), 5, time.Second, 1, func(context.Context) error {
+		_, err = sp.db.Exec(sql)
+		if err != nil {
+			return errors.Annotatef(err, "query sql failed: %s", sql)
+		}
+		return nil
+	})
 }
 
 // IsConsistent implements CheckPoint interface

--- a/drainer/checkpoint/mysql_test.go
+++ b/drainer/checkpoint/mysql_test.go
@@ -50,7 +50,7 @@ func (s *saveSuite) TestShouldSaveCheckpoint(c *C) {
 	c.Assert(err, IsNil)
 	mock.ExpectExec("replace into db.tbl.*").WillReturnResult(sqlmock.NewResult(0, 0))
 	cp := MysqlCheckPoint{db: db, schema: "db", table: "tbl"}
-	err = cp.Save(1111, 0, false)
+	err = cp.Save(1111, 0, false, 0)
 	c.Assert(err, IsNil)
 }
 
@@ -64,7 +64,7 @@ func (s *saveSuite) TestShouldUpdateTsMap(c *C) {
 		table:  "tbl",
 		TsMap:  make(map[string]int64),
 	}
-	err = cp.Save(65536, 3333, false)
+	err = cp.Save(65536, 3333, false, 0)
 	c.Assert(err, IsNil)
 	c.Assert(cp.TsMap["primary-ts"], Equals, int64(65536))
 	c.Assert(cp.TsMap["secondary-ts"], Equals, int64(3333))

--- a/drainer/relay.go
+++ b/drainer/relay.go
@@ -150,7 +150,7 @@ func feedByRelayLog(r relay.Reader, ld loader.Loader, cp checkpoint.CheckPoint) 
 		return errors.Trace(readerErr)
 	}
 
-	err := cp.Save(lastSuccessTS, 0 /* secondaryTS */, true /*consistent*/)
+	err := cp.Save(lastSuccessTS, 0 /* secondaryTS */, true /*consistent*/, 0)
 	if err != nil {
 		return errors.Trace(err)
 	}

--- a/drainer/relay_test.go
+++ b/drainer/relay_test.go
@@ -74,7 +74,7 @@ var _ loader.Loader = &noOpLoader{}
 func (s *relaySuite) TestFeedByRealyLog(c *check.C) {
 	cp, err := checkpoint.NewFile(0 /* initialCommitTS */, c.MkDir()+"/checkpoint")
 	c.Assert(err, check.IsNil)
-	err = cp.Save(0, 0, false)
+	err = cp.Save(0, 0, false, 0)
 	c.Assert(err, check.IsNil)
 	c.Assert(cp.IsConsistent(), check.Equals, false)
 

--- a/drainer/sync/syncer.go
+++ b/drainer/sync/syncer.go
@@ -31,7 +31,7 @@ type Item struct {
 	// Each item has a schemaVersion. with amend txn feature the prewrite DML's SchemaVersion could change.
 	// which makes restart & reload history DDL with previous SchemaVersion not reliable.
 	// so we should save this version as checkpoint.
-	SchemaVersion   int64
+	SchemaVersion int64
 
 	// the applied TS executed in downstream, only for tidb
 	AppliedTS int64

--- a/drainer/sync/syncer.go
+++ b/drainer/sync/syncer.go
@@ -28,6 +28,11 @@ type Item struct {
 	Table         string
 	RelayLogPos   pb.Pos
 
+	// Each item has a schemaVersion. with amend txn feature the prewrite DML's SchemaVersion could change.
+	// which makes restart & reload history DDL with previous SchemaVersion not reliable.
+	// so we should save this version as checkpoint.
+	SchemaVersion   int64
+
 	// the applied TS executed in downstream, only for tidb
 	AppliedTS int64
 	// should skip to replicate this item at downstream

--- a/drainer/syncer.go
+++ b/drainer/syncer.go
@@ -184,6 +184,7 @@ func (s *Syncer) enableSafeModeInitializationPhase() {
 func (s *Syncer) handleSuccess(fakeBinlog chan *pb.Binlog, lastTS *int64) {
 	successes := s.dsyncer.Successes()
 	var lastSaveTS int64
+	var latestVersion int64
 	lastSaveTime := time.Now()
 
 	for {
@@ -208,6 +209,7 @@ func (s *Syncer) handleSuccess(fakeBinlog chan *pb.Binlog, lastTS *int64) {
 			if ts > atomic.LoadInt64(lastTS) {
 				atomic.StoreInt64(lastTS, ts)
 			}
+			latestVersion = item.SchemaVersion
 
 			// save ASAP for DDL, and if FinishTS > 0, we should save the ts map
 			if item.Binlog.DdlJobId > 0 || item.AppliedTS > 0 {
@@ -229,7 +231,7 @@ func (s *Syncer) handleSuccess(fakeBinlog chan *pb.Binlog, lastTS *int64) {
 		ts := atomic.LoadInt64(lastTS)
 		if ts > lastSaveTS {
 			if saveNow || time.Since(lastSaveTime) > 3*time.Second {
-				s.savePoint(ts, appliedTS)
+				s.savePoint(ts, appliedTS, latestVersion)
 				lastSaveTime = time.Now()
 				lastSaveTS = ts
 				appliedTS = 0
@@ -242,20 +244,20 @@ func (s *Syncer) handleSuccess(fakeBinlog chan *pb.Binlog, lastTS *int64) {
 
 	ts := atomic.LoadInt64(lastTS)
 	if ts > lastSaveTS {
-		s.savePoint(ts, 0)
+		s.savePoint(ts, 0, latestVersion)
 		eventCounter.WithLabelValues("savepoint").Add(1)
 	}
 
 	log.Info("handleSuccess quit")
 }
 
-func (s *Syncer) savePoint(ts, secondaryTS int64) {
+func (s *Syncer) savePoint(ts, secondaryTS, version int64) {
 	if ts < s.cp.TS() {
 		log.Error("save ts is less than checkpoint ts %d", zap.Int64("save ts", ts), zap.Int64("checkpoint ts", s.cp.TS()))
 	}
 
 	log.Info("write save point", zap.Int64("ts", ts))
-	err := s.cp.Save(ts, secondaryTS, false)
+	err := s.cp.Save(ts, secondaryTS, false, version)
 	if err != nil {
 		log.Fatal("save checkpoint failed", zap.Int64("ts", ts), zap.Error(err))
 	}
@@ -281,6 +283,12 @@ func (s *Syncer) run() error {
 	var err error
 
 	s.enableSafeModeInitializationPhase()
+
+	err = s.schema.handlePreviousDDLJobIfNeed(s.cp.SchemaVersion())
+	if err != nil {
+		err = errors.Annotate(err, "handlePreviousDDLJobIfNeed failed")
+		return err
+	}
 
 	var lastDDLSchemaVersion int64
 	var b *binlogItem
@@ -354,6 +362,7 @@ ForLoop:
 				err = errors.Annotate(err, "handlePreviousDDLJobIfNeed failed")
 				break ForLoop
 			}
+
 			var isFilterTransaction = false
 			var err1 error
 			if s.loopbackSync != nil && s.loopbackSync.LoopbackControl {
@@ -375,7 +384,7 @@ ForLoop:
 				s.addDMLEventMetrics(preWrite.GetMutations())
 				beginTime := time.Now()
 				lastAddComitTS = binlog.GetCommitTs()
-				err = s.dsyncer.Sync(&dsync.Item{Binlog: binlog, PrewriteValue: preWrite})
+				err = s.dsyncer.Sync(&dsync.Item{Binlog: binlog, PrewriteValue: preWrite, SchemaVersion: preWrite.SchemaVersion})
 				if err != nil {
 					err = errors.Annotatef(err, "failed to add item")
 					break ForLoop
@@ -396,7 +405,6 @@ ForLoop:
 
 			log.Debug("get DDL", zap.Int64("SchemaVersion", b.job.BinlogInfo.SchemaVersion))
 			lastDDLSchemaVersion = b.job.BinlogInfo.SchemaVersion
-
 			err = s.schema.handlePreviousDDLJobIfNeed(b.job.BinlogInfo.SchemaVersion)
 			if err != nil {
 				err = errors.Trace(err)
@@ -443,7 +451,7 @@ ForLoop:
 			log.Info("add ddl item to syncer, you can add this commit ts to `ignore-txn-commit-ts` to skip this ddl if needed",
 				zap.String("sql", sql), zap.Int64("commit ts", binlog.CommitTs))
 
-			err = s.dsyncer.Sync(&dsync.Item{Binlog: binlog, PrewriteValue: nil, Schema: schema, Table: table, ShouldSkip: shouldSkip})
+			err = s.dsyncer.Sync(&dsync.Item{Binlog: binlog, PrewriteValue: nil, Schema: schema, Table: table, ShouldSkip: shouldSkip, SchemaVersion: lastDDLSchemaVersion})
 			if err != nil {
 				err = errors.Annotatef(err, "add to dsyncer, commit ts %d", binlog.CommitTs)
 				break ForLoop
@@ -473,7 +481,7 @@ ForLoop:
 		return cerr
 	}
 
-	return s.cp.Save(s.cp.TS(), 0, true /*consistent*/)
+	return s.cp.Save(s.cp.TS(), 0, true /*consistent*/, lastDDLSchemaVersion)
 }
 
 func findLoopBackMark(dmls []*loader.DML, info *loopbacksync.LoopBackSync) (bool, error) {

--- a/drainer/syncer.go
+++ b/drainer/syncer.go
@@ -256,7 +256,7 @@ func (s *Syncer) savePoint(ts, secondaryTS, version int64) {
 		log.Error("save ts is less than checkpoint ts %d", zap.Int64("save ts", ts), zap.Int64("checkpoint ts", s.cp.TS()))
 	}
 
-	log.Info("write save point", zap.Int64("ts", ts))
+	log.Info("write save point", zap.Int64("ts", ts), zap.Int64("version", version))
 	err := s.cp.Save(ts, secondaryTS, false, version)
 	if err != nil {
 		log.Fatal("save checkpoint failed", zap.Int64("ts", ts), zap.Error(err))

--- a/drainer/syncer.go
+++ b/drainer/syncer.go
@@ -259,7 +259,7 @@ func (s *Syncer) savePoint(ts, secondaryTS, version int64) {
 	log.Info("write save point", zap.Int64("ts", ts), zap.Int64("version", version))
 	err := s.cp.Save(ts, secondaryTS, false, version)
 	if err != nil {
-		log.Error("save checkpoint failed", zap.Int64("ts", ts), zap.Int64("version", version), zap.Error(err))
+		log.Fatal("save checkpoint failed", zap.Int64("ts", ts), zap.Int64("version", version), zap.Error(err))
 	}
 
 	checkpointTSOGauge.Set(float64(oracle.ExtractPhysical(uint64(ts))))

--- a/drainer/syncer.go
+++ b/drainer/syncer.go
@@ -259,7 +259,7 @@ func (s *Syncer) savePoint(ts, secondaryTS, version int64) {
 	log.Info("write save point", zap.Int64("ts", ts), zap.Int64("version", version))
 	err := s.cp.Save(ts, secondaryTS, false, version)
 	if err != nil {
-		log.Fatal("save checkpoint failed", zap.Int64("ts", ts), zap.Error(err))
+		log.Error("save checkpoint failed", zap.Int64("ts", ts), zap.Int64("version", version), zap.Error(err))
 	}
 
 	checkpointTSOGauge.Set(float64(oracle.ExtractPhysical(uint64(ts))))

--- a/drainer/syncer.go
+++ b/drainer/syncer.go
@@ -284,7 +284,7 @@ func (s *Syncer) run() error {
 
 	s.enableSafeModeInitializationPhase()
 
-	err = s.schema.handlePreviousDDLJobIfNeed(s.cp.SchemaVersion())
+	err = s.schema.handlePreviousDDLJobIfNeed(s.cp.SchemaVersion()+1)
 	if err != nil {
 		err = errors.Annotate(err, "handlePreviousDDLJobIfNeed failed")
 		return err

